backend/internal/compile/resolve/http.go

```diff
@@ -34,68 +28,79 @@ func NewHttpResolver(givenUrl *url.URL) *Resolver {
 	return &Resolver{
 		FS: &HttpResolverFs{
 			BaseURL: baseUrl,
+			client:  client,
 		},
 	}
 }
 
 // HttpFileEntry is an implementation of fs.File that is backed by an HTTP response.
-// This lines up with assumption (2) above.
 type HttpFileEntry struct {
-	res *http.Response
+	path     string
+	contents string
 }
 
 func (entry HttpFileEntry) Name() string {
-	return path.Base(entry.res.Request.URL.Path)
+	return path.Base(entry.path)
 }
 
 func (entry HttpFileEntry) Size() int64 {
-	return entry.res.ContentLength
+	panic(fmt.Errorf("unsupported method"))
 }
 
 func (entry HttpFileEntry) Mode() fs.FileMode {
-	return 0
+	panic(fmt.Errorf("unsupported method"))
 }
 
 func (entry HttpFileEntry) ModTime() time.Time {
-	return time.Now()
+	panic(fmt.Errorf("unsupported method"))
 }
 
 func (entry HttpFileEntry) IsDir() bool {
-	return false
+	panic(fmt.Errorf("unsupported method"))
 }
 
 func (entry HttpFileEntry) Sys() interface{} {
-	return nil
+	panic(fmt.Errorf("unsupported method"))
 }
 
 func (entry HttpFileEntry) Stat() (fs.FileInfo, error) {
 	return entry, nil
 }
 
 func (entry HttpFileEntry) Read(p []byte) (n int, err error) {
-	return entry.res.Body.Read(p)
+	n = copy(p, entry.contents)
+	return n, io.EOF
 }
 
 func (entry HttpFileEntry) Close() error {
-	return entry.res.Body.Close()
+	return nil
 }
 
 func (hfs *HttpResolverFs) Open(filename string) (fs.File, error) {
 	fileUrl := hfs.BaseURL.ResolveReference(&url.URL{Path: filename})
-	resp, err := http.Get(fileUrl.String())
-	if err != nil {
-		return nil, err
-	}
-	if resp.StatusCode == http.StatusNotFound {
-		resp.Body.Close()
-		return nil, fmt.Errorf("file not found: %s (status %d)", filename, resp.StatusCode)
-	}
-	if resp.StatusCode != http.StatusOK {
-		buf, err := io.ReadAll(resp.Body)
+
+	// This is a bit silly, but esm.sh is really slow to detect bad URLs, but unpkg.com is very
+	// fast thanks to edge caching. So we check if the file exists on unpkg.com first, and if it
+	// does, we allow the request to go through.
+	//
+	// unpkg.com also realizes that the URL is actually immutable, and will ask the client to cache
+	// it while esm.sh reports the response as 'no-cache'.
+	if fileUrl.Scheme == "https" && fileUrl.Host == "esm.sh" {
```

Comment:
Benchmark without the hack:

<img width="1181" alt="Screenshot 2023-03-01 at 7 13 50 PM" src="https://user-images.githubusercontent.com/2664912/222296529-7d6c03a4-fda2-42ff-a64d-2571439f5b9b.png">

Benchmark with the hack:

<img width="1181" alt="Screenshot 2023-03-01 at 7 14 11 PM" src="https://user-images.githubusercontent.com/2664912/222296609-5eedc14f-a440-4a5f-b3a3-ba43879a6f88.png">